### ISSUE TITLE
Storing appliance fingerprints/signatures/priors needs definition

### ISSUE THREAD

**Issue by gjwo** [nilmtk/nilmtk#413]
NILMTK doesn't appear to have an agreed format for the storage of a set of labelled fingerprints for appliances, i.e. a way of decoupling the training and disaggregation steps. Have any "standard" methods come into common use? If not the following elements would seem to be needed
*  **Result label** (mandatory) <String> the name label associated with appliance described
*  **Result metadata** (optional) further information on the appliance, such as  location - building, room, meter; list of components; see appliance in nilmtk_metadata
*  **Signature** (mandatory) <DataFrame> The signature the appliance will produce in the data - there are a variety of ways this could be represented in a dataframe
*  **Signature metadata** (optional) a description of the signature characteristics, such as granularity, compression methods etc. NB a basic signature type may in fact be mandatory if there can be more than one type

This would need to be stored in some kind of defined file format possibly HDF5 so that it could be loaded at a late point in time and used by a disaggregation algorithm.

**Comment by JackKelly** [comment]
I must admit I don't fully understand what you count as a 'a set of labelled fingerprints for appliances'.

NILMTK already has a way to store multiple submeter streams (e.g. for the fridge, tv, kettle etc).  Is this what you mean?  A typical NILM algorithm will produce data which is in an identical format to 'real' submetered data; i.e. a 2D matrix where one axis is time and the other axis is each appliance.

**Comment by gjwo** [comment]
Probably not the same thing, ignoring for a minute constantly cycling appliances like fridges, the fingerprint for a washing machine might be:
*  result label: 'Washing Machine1' 
*  metadata: as per the building metadata for my washing machine, make model etc
*  Signature: A time series dataframe staring from time of the Real / Apparent / Reactive readings from the washing machine in question, or some more abstract representation of this such as the expected features.
* Signature metadata: type: <as submetered> or <abstract type 1> etc, Repeat information, probability of use at specific times etc. Variations in on cycles and off cycles etc

If I have understood you correctly are you  talking about a time series with a column for the activity of each appliance on the submeter? An example would help.

It should be trivial to produce this for a house with full metadata, for a sub-metered appliance with the an 'on period' identified, The fingerprint would then be used via clever correlation to identify all the other times that device was on without the benefit of a sub-meter. A single sub meter could be used to sequentially capture and store the fingerprints of each of the major appliances in a building (tedious but fairly straight forward). 

This data should correspond to some of the public appliance databases Ollie mentioned in his blog.

**Comment by JackKelly** [comment]
Our HDF files contain a DataFrame (2D matrix) for each meter.  Each DataFrame contains every sample recorded by that meter.  So if, as is the case with pretty much every dataset we're interested in, you have multiple submeters per house then you'll get simultaneous, timestamped recordings of each submeter and of the whole-house aggregate.  As far as I can tell, our HDF files already contain the first three items on your list above.  Many of the items in your final bullet point can be calculated from the data by NILMTK's statistics algorithms.

Maybe have a look at my [UK-DALE](http://www.doc.ic.ac.uk/~dk3810/data/) paper for more details of my dataset, as an example (or Nipun's dataset, or REDD, or any dataset really).  And the HDF file contains all the metadata for the appliances, submeters, buildings etc.

There's also [`Electric::activation_series()`](http://nilmtk.github.io/nilmtk/master/nilmtk.html#nilmtk.electric.Electric.activation_series) which can identify individual 'activations' of an appliance and returns a list of `pd.Series`, one for each appliance activation.

**Comment by gjwo** [comment]
@JackKelly @nipunreddevil  (as you are looking at Hart85 soon) 
After having explored the implementation of Hart in nilmtk at some length, and having read Hart's '85 and '92 papers, I think I have a better handle on this now.  See summary of Hart's '92 process below
![image](https://cloud.githubusercontent.com/assets/12109990/10513090/0b817b4c-733d-11e5-998c-79068d627f9e.png)

The result of the cluster analysis in nilmtk terms is the list of centroids that have been build by the training step. A single centroid is a rudimentary model for an on/off type of appliance,  by extension a list of centroids would be a rudimentary model for a multi-state or FSM type of appliance (disregarding any continuously variable states or appliances).

One of the things I believe is currently missing from the Hart implementation in nilmtk is step 'H' above i.e. the linking of the centroid models to any appliance metadata. If the typical steady power states (active and reactive) were recorded in the metadata, some automatic matching of names to observed patterns would be possible following training and could then be used for disaggregation. This initial centroid values would need to be entered in metadata manually from a training run to bootstrap this process as the plate values for power consumption are unlikely to be accurate enough. There would also need to be a definition of how far from a centroid a result would be an acceptable match.  

The results from training could be stored as unnamed centroids, along with metadata including ALL parameters to the training step. The linkage of names to centroids, could be (optionally) stored in the same dataset if the linkage were available. "Unknown n" could be used for any centroids not linked.

It would them be possible to produce the kind of appliance usage charts included in Hart's papers.

Appliance Signatures can clearly be more complex than this (see taxonomy below from the '92 paper
![image](https://cloud.githubusercontent.com/assets/12109990/10513598/a276e372-7340-11e5-82cc-b3bafb5317d8.png)
In particular the transients could be of use in further differentiating between devices.

**Comment by JackKelly** [comment]
> One of the things I believe is currently missing from the Hart implementation in nilmtk is step 'H' above i.e. the linking of the centroid models to any appliance metadata. If the typical steady power states (active and reactive) were recorded in the metadata, some automatic matching of names to observed patterns would be possible following training and could then be used for disaggregation. This initial centroid values would need to be entered in metadata manually from a training run to bootstrap this process as the plate values for power consumption are unlikely to be accurate enough. There would also need to be a definition of how far from a centroid a result would be an acceptable match.
> 
> The results from training could be stored as unnamed centroids, along with metadata including ALL parameters to the training step. The linkage of names to centroids, could be (optionally) stored in the same dataset if the linkage were available. "Unknown n" could be used for any centroids not linked.
> 
> It would them be possible to produce the kind of appliance usage charts included in Hart's papers.

These are all excellent ideas and, I agree, NILMTK should be able to do this.

To be honest, I'm not sure I'll be able to justify spending the time to do this work, though.  My research focus is very much on the application of neural nets to NILM at the moment, and probably will be for the foreseeable future (although, for the next few months my focus is very much of writing my thesis rather than writing code, unfortunately!).  Maybe one of my MSc / BSc students might be interested in doing this.

Of course, it would be amazing if you wanted to have a shot at implementing this!  It sounds like you've taught yourself a lot from Hart's papers!  We'll give you as much help as we can.

**Comment by gjwo** [comment]
While I remember - For devices with sub components in the metadata, the list of centroids could be derived by a tree walk of the sub components if the centroids were stored in them. Any initial pass would show a bunch of components, which could then be associated into the complex appliance in a second pass.

I am working up to doing some of this, the first step is to get it straight in my head, and sanity check it with you guys. The next step is getting more comfortable with python and pandas, which may take longer, as I find a lot of the ways Python works is at odds with my programming background

**Comment by PMeira** [comment]
I should handle something like this in the coming weeks/months.
